File: 78-Subsets/78-Subsets.java
// Last updated: 16/10/2025, 04:29:57
class Solution {

    public void helper(int idx, int n, int[] nums, List<Integer> ds, List<List<Integer>> ans){
        if(idx >= n){
            ans.add(new ArrayList<>(ds));
            return;
        }
        ds.add(nums[idx]);
        helper(idx + 1, n, nums, ds, ans); //pick
        ds.remove(ds.size() - 1);
        helper(idx + 1, n, nums, ds, ans); //not pick
    }

    public List<List<Integer>> subsets(int[] nums) {
        int n = nums.length;
        List<Integer> ds = new ArrayList<Integer>();
        List<List<Integer>> ans = new ArrayList<>();
        helper(0, n, nums, ds, ans);
        return ans;
    }
}

File: 78-Subsets/78-Subsets.cpp
// Last updated: 16/10/2025, 04:30:02
#include<vector>
class Solution {
    
public:
    vector<vector<int>> ans;
    void  helper(int ind, vector<int> &ds, vector<int> &nums, int n){
        if(ind == n){
            ans.push_back(ds);
            return;
        }
        ds.push_back(nums[ind]);
        helper(ind+1,ds,nums,n);
        ds.pop_back();
        helper(ind+1,ds,nums,n);
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        vector<int> ds;
        helper(0,ds,nums,n);
        return ans;
    }
};